// fh-amqp-js/lib/fhevents.js
// App related events
exports.DYNOMAN_APP_START_FAILED = "fh.events.dynoman.app.start-failed";
exports.DYNOMAN_APP_START_SUCCEEDED = "fh.events.dynoman.app.start-succeeded";
exports.DYNOMAN_APP_STOP_FAILED = "fh.events.dynoman.app.stop-failed";
exports.DYNOMAN_APP_STOP_SUCCEEDED = "fh.events.dynoman.app.stop-succeeded";
exports.DYNOMAN_APP_STAGE_FAILED = "fh.events.dynoman.app.stage-failed";
exports.DYNOMAN_APP_STAGE_SUCCEEDED = "fh.events.dynoman.app.stage-succeeded";
exports.DYNOMAN_APP_DELETE_FAILED = "fh.events.dynoman.app.delete-failed";
exports.DYNOMAN_APP_DELETE_SUCCEEDED = "fh.events.dynoman.app.delete-succeeded";
exports.DYNOMAN_APP_SUSPEND_SUCCEEDED = "fh.events.dynoman.app.suspend-succeeded";
exports.DYNOMAN_APP_IMPORT_SUCCEEDED = "fh.events.dynoman.app.import-succeeded";
exports.DYNOMAN_APP_EXPORT_SUCCEEDED = "fh.events.dynoman.app.export-succeeded";

// Millicore related events
exports.CORE_APP_START_REQUESTED = "fh.events.millicore.app.start-reqeusted";
exports.CORE_APP_STOP_REQUESTED = "fh.events.millicore.app.stop-requested";
exports.CORE_APP_DEPLOY_REQUESTED = "fh.events.millicore.app.deploy-requested";
exports.CORE_APP_ENV_CHANGE_REQUESTED = "fh.events.millicore.env.change-requested";
exports.CORE_APP_CREATE_REQUESTED = "fh.events.millicore.app.create-requested";
exports.CORE_APP_DELETE_REQUESTED = "fh.events.millicore.app.delete-requested";
// TODO - revisit when we re-support CF
/*
exports.CORE_CF_APP_DEPLOYED = "fh.events.millicore.app.cf.deploy-succeeded";
exports.CORE_CF_APP_DEPLOY_FAILED = "fh.events.millicore.app.cf.deploy-failed";
exports.CORE_CF_APP_STARTED = " fh.events.millicore.app.cf.start-succeeded";
exports.CORE_CF_APP_START_FAILED = "fh.events.millicore.app.cf.start-failed";
exports.CORE_CF_APP_STOPPED = "fh.events.millicore.app.cf.stop-succeeded";
exports.CORE_CF_APP_STOP_FAILED = " fh.events.millicore.app.cf.stop-failed";
*/

// Dyno related events
exports.DYNOMAN_DYNO_CREATE_SUCCEEDED = "fh.events.dynoman.dyno.create-succeeded";
exports.DYNOMAN_DYNO_DELETE_SUCCEEDED = "fh.events.dynoman.dyno.delete-succeeded";
exports.DYNOMAN_DYNO_START_SUCCEEDED  = "fh.events.dynoman.dyno.start-succeeded";
exports.DYNOMAN_DYNO_STOP_SUCCEEDED  = "fh.events.dynoman.dyno.stop-succeeded";
exports.DYNOMAN_RESOURCES_CHANGED_SUCCEEDED = "fh.events.dynoman.dyno.resources-changed-succeeded";
exports.DYNOMAN_DYNO_IMPORT_SUCCEEDED  = "fh.events.dynoman.dyno.import-succeeded";
exports.DYNOMAN_DYNO_EXPORT_SUCCEEDED  = "fh.events.dynoman.dyno.export-succeeded";

// Monit related events
exports.MONIT_APP_TERMINATED = "fh.events.monit.app.terminated";
exports.MONIT_APP_HIGH_CPU = "fh.events.monit.app.high-cpu";
exports.MONIT_APP_HIGH_MEMORY = "fh.events.monit.app.high-memory";

var core = {};
core[exports.CORE_APP_START_REQUESTED] = {
  "eventClass": "APP_STATE",
  "eventType": "START_REQUESTED",
  "eventLevel": "INFO"
};

core[exports.CORE_APP_STOP_REQUESTED] = {
  "eventClass": "APP_STATE",
  "eventType": "STOP_REQUESTED",
  "eventLevel": "INFO"
};

core[exports.CORE_APP_DEPLOY_REQUESTED] = {
  "eventClass": "APP_STATE",
  "eventType": "DEPLOY_REQUESTED",
  "eventLevel": "INFO"
};

core[exports.CORE_APP_ENV_CHANGE_REQUESTED] = {
  "eventClass": "APP_ENVIRONMENT",
  "eventType": "CHANGE_REQUESTED",
  "eventLevel": "INFO"
};

core[exports.CORE_APP_CREATE_REQUESTED] = {
  "eventClass": "APP_STATE",
  "eventType": "CREATE_REQUESTED",
  "eventLevel": "INFO"
};

core[exports.CORE_APP_DELETE_REQUESTED] = {
  "eventClass": "APP_STATE",
  "eventType": "DELETE_REQUESTED",
  "eventLevel": "INFO"
};

/* TODO - revisit when we re-support CF
 *

core[exports.CORE_CF_APP_DEPLOYED] = {
  "eventClass": "APP_STATE",
  "eventType": "DEPLOYED",
  "eventLevel": "INFO"
};

core[exports.CORE_CF_APP_DEPLOY_FAILED] = {
  "eventClass": "exports.CORE_CF_APP_DEPLOY_FAILED",
  "eventType": exports.CORE_CF_APP_DEPLOY_FAILED,
  "eventLevel": "INFO"
};

core[exports.CORE_CF_APP_STARTED] = {
  "eventClass": "exports.CORE_CF_APP_STARTED",
  "eventType": exports.CORE_CF_APP_STARTED,
  "eventLevel": "INFO"
};

core[exports.CORE_CF_APP_START_FAILED] = {
  "eventClass": "exports.CORE_CF_APP_START_FAILED",
  "eventType": exports.CORE_CF_APP_START_FAILED,
  "eventLevel": "INFO"
};

core[exports.CORE_CF_APP_STOPPED] = {
  "eventClass": "exports.CORE_CF_APP_STOPPED",
  "eventType": exports.CORE_CF_APP_STOPPED,
  "eventLevel": "INFO"
};

core[exports.CORE_CF_APP_STOP_FAILED] = {
  "eventClass": "exports.CORE_CF_APP_STOP_FAILED",
  "eventType": exports.CORE_CF_APP_STOP_FAILED,
  "eventLevel": "INFO"
};
*/

var dynoman = {};
dynoman[exports.DYNOMAN_APP_START_FAILED] = {
  "eventClass":"APP_STATE",
  "eventType":"START_FAILED",
  "eventLevel":"ERROR"
};

dynoman[exports.DYNOMAN_APP_START_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"START_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_STOP_FAILED] = {
  "eventClass":"APP_STATE",
  "eventType":"START_FAILED",
  "eventLevel":"ERROR"
};

dynoman[exports.DYNOMAN_APP_STOP_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"STOP_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_SUSPEND_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"SUSPEND_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_STAGE_FAILED] = {
  "eventClass":"APP_STATE",
  "eventType":"DEPLOY_FAILED",
  "eventLevel":"ERROR"
};

dynoman[exports.DYNOMAN_APP_STAGE_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"DEPLOYED",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_DELETE_FAILED] = {
  "eventClass":"APP_STATE",
  "eventType":"DELETED",
  "eventLevel":"ERROR"
};

dynoman[exports.DYNOMAN_APP_DELETE_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"DELETED",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_IMPORT_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"IMPORTED",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_APP_EXPORTED_SUCCEEDED] = {
  "eventClass":"APP_STATE",
  "eventType":"EXPORTED",
  "eventLevel":"INFO"
};

// Dyno Related event details
dynoman[exports.DYNOMAN_DYNO_CREATE_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"CREATED",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_DYNO_DELETE_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"DELETED",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_DYNO_START_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"START_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_DYNO_STOP_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"STOP_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_RESOURCES_CHANGED_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"RESOURCES_CHANGED_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_DYNO_IMPORT_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"IMPORT_SUCCESSFUL",
  "eventLevel":"INFO"
};

dynoman[exports.DYNOMAN_DYNO_EXPORT_SUCCEEDED] = {
  "eventClass":"DYNO_STATE",
  "eventType":"EXPORT_SUCCESSFUL",
  "eventLevel":"INFO"
};

// Monit related event details
var monit = {};

monit[exports.MONIT_APP_TERMINATED] = {
  "eventClass":"APP_STATE",
  "eventType":"TERMINATED",
  "eventLevel":"FATAL",
  "eventMessage":"App failed to restart - %s"
};

monit[exports.MONIT_APP_HIGH_CPU] = {
  "eventClass": "APP_STATE",
  "eventType":"HIGH_CPU",
  "eventLevel": "WARN",
  "eventMessage": "App has high CPU usage - %s"
};

monit[exports.MONIT_APP_HIGH_MEMORY] = {
  "eventClass": "APP_STATE",
  "eventType":"HIGH_MEMORY",
  "eventLevel": "WARN",
  "eventMessage": "App has high memroy usage - %s"
};

exports.dynoman = dynoman;
exports.monit = monit;
exports.core = core;
